python/dolfinx/wrappers/graph.cpp

```diff
@@ -88,32 +88,34 @@ void declare_adjacency_list(nb::module_& m, std::string type)
           "Links (edges) of a node")
       .def_prop_ro(
           "array",
-          [](const dolfinx::graph::AdjacencyList<T>& self)
+          [](const dolfinx::graph::AdjacencyList<T, U>& self)
           {
             return nb::ndarray<const T, nb::numpy>(self.array().data(),
                                                    {self.array().size()});
           },
           nb::rv_policy::reference_internal)
       .def_prop_ro(
           "offsets",
-          [](const dolfinx::graph::AdjacencyList<T>& self)
+          [](const dolfinx::graph::AdjacencyList<T, U>& self)
           {
             return nb::ndarray<const std::int32_t, nb::numpy>(
                 self.offsets().data(), {self.offsets().size()});
           },
           nb::rv_policy::reference_internal)
-      .def_prop_ro("num_nodes", &dolfinx::graph::AdjacencyList<T>::num_nodes)
-      .def("__eq__", &dolfinx::graph::AdjacencyList<T>::operator==,
+      .def_prop_ro("num_nodes", &dolfinx::graph::AdjacencyList<T, U>::num_nodes)
+      .def("__eq__", &dolfinx::graph::AdjacencyList<T, U>::operator==,
            nb::is_operator())
-      .def("__repr__", &dolfinx::graph::AdjacencyList<T>::str)
-      .def("__len__", &dolfinx::graph::AdjacencyList<T>::num_nodes);
+      .def("__repr__", &dolfinx::graph::AdjacencyList<T, U>::str)
+      .def("__len__", &dolfinx::graph::AdjacencyList<T, U>::num_nodes);
 }
 
 void graph(nb::module_& m)
 {
-
-  declare_adjacency_list<std::int32_t>(m, "int32");
-  declare_adjacency_list<std::int64_t>(m, "int64");
+  declare_adjacency_list<std::int32_t, std::nullptr_t>(m, "int32");
+  declare_adjacency_list<std::int64_t, std::nullptr_t>(m, "int64");
+  nb::class_<dolfinx::graph::AdjacencyList<
+      std::tuple<int, std::size_t, std::int8_t>, std::int32_t>>(
+      m, "Adjacency_List_int_size_t_int8__int32", "Adjacency List");
```

Comment:
```suggestion
      m, "Adjacency_List_int_size_t_int8_int32", "Adjacency List");
```
Just a typo.

Furthermore, as you pointed out in https://github.com/FEniCS/dolfinx/pull/3108#discussion_r1536660501, how far are these templates going to go in the python bindings?

Reply:
As far as the type that we use from Python.